pkg/syncserver/sync_server.go

```diff
@@ -199,20 +226,43 @@ func (s *Server) SetMaxConns(numConns int) {
 	s.maxConnsC <- numConns
 }
 
+func (s *Server) Port() int {
+	<-s.listeningC
+	return s.chosenPort
+}
+
 func (s *Server) serve(cxt context.Context) {
-	logCxt := log.WithField("port", syncproto.DefaultPort)
+	defer s.Finished.Done()
+	var cancelFn context.CancelFunc
+	cxt, cancelFn = context.WithCancel(cxt)
+	defer cancelFn()
+
+	logCxt := log.WithField("port", s.config.Port)
 	logCxt.Info("Opening listen socket")
-	l, err := net.ListenTCP("tcp", &net.TCPAddr{Port: syncproto.DefaultPort})
+	l, err := net.ListenTCP("tcp", &net.TCPAddr{Port: s.config.ListenPort()})
 	if err != nil {
 		logCxt.WithError(err).Panic("Failed to open listen socket")
 	}
 	logCxt.Info("Opened listen socket")
-	defer l.Close()
+	s.Finished.Add(1)
+	go func() {
+		<-cxt.Done()
+		l.Close()
+		s.Finished.Done()
+	}()
+	s.chosenPort = l.Addr().(*net.TCPAddr).Port
+	logCxt = log.WithField("port", s.chosenPort)
+	close(s.listeningC)
```

Comment:
You're absolutely right in general, but the close channel/receive zero value pattern I used there establishes a "happens before" relationship between the code that runs before the close in the closing goroutine (i.e. the write to the variable) and the code that runs after the read in the reading goroutine (so go's memory model makes the guarantee that the reading goroutine will see the result of the closing goroutine's write to the port).

The compiler is free to implement that will a write barrier or some other primitive.